### PR TITLE
[SYCL] Implement USM vars and placeholder accessors passed to reduction

### PR DESCRIPTION
**Pull request by v-klochkov**
In order to support USM vars/pointers passed to reduction
the placeholder accessors to global buffers were used.

Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/intel/reduction.hpp [hunk 1]
```diff
@@ -408,14 +470,30 @@ class reduction_impl {
...
   }
 
-  /// User's accessor to where the reduction must be written.
-  accessor_type MAcc;
+  accessor_type &getUserAccessor() { return MAcc; }
+
+  T *getUSMPointer() {
+    assert(is_usm && "Unexpected call of getUSMPointer().");
+    return MUSMPointer;
+  }
```
Comment (Pennycook) [inline sycl/include/CL/sycl/intel/reduction.hpp:478]: This is a question, not a request for a change.  Would it make sense to use SFINAE to disable these in cases that they won't work, or do you think this is not worth the pain (because users will never call these functions anyway)?

File: sycl/test/reduction/reduction_usm.cpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+template <typename T, int Dim, class BinaryOperation>
+void test(T Identity, size_t WGSize, size_t NWItems) {
+  // Initialize.
+  T CorrectOut;
+  BinaryOperation BOp;
+
+  buffer<T, 1> InBuf(NWItems);
+  initInputData(InBuf, CorrectOut, Identity, BOp, NWItems);
+
+  T ReduVar = Identity;
```
Comment (Pennycook) [inline sycl/test/reduction/reduction_usm.cpp:31]: `ReduVar` isn't a USM allocation.

I like what you've implemented, but it isn't performing reductions using a variable allocated in USM -- it's performing reductions using a variable allocated on the host.  This may be an important use-case, and I wonder if we should update the reduction proposal to make it legal... any thoughts, @rolandschulz @jasonsewall-intel?

The use-case we were targeting with USM would look something like the following:
```c++
queue Q;
T* ReduVar = malloc_device(sizeof(T), q);
Q.submit([&](handler &CGH) {
  auto Redu = intel::reduction(ReduVar, Identity, BOp);
  ...
});
```

File: sycl/include/CL/sycl/intel/reduction.hpp (hunk 1 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/intel/reduction.hpp:478]: I thought about that, but decided not adding SFINAE for this method now because

a) reduction_impl lives on HOST and it is not such a big overhead for it to carry MUSMPointer field and have this method also. Also, user is not supposed to call this method, it is closed.

b) adding SFINAE to getUSMPointer will require to have split parallel_for() at handler.hpp:854 because I wouldn't be able to have that simple if handler.hpp:863.
Similar for another parallel_for using that method (handler.hpp:983)

So, I thought it is more code without clear benefit.

File: sycl/test/reduction/reduction_usm.cpp (hunk 2 above)
Comment (v-klochkov) [reply sycl/test/reduction/reduction_usm.cpp:31]: Good catch, I misunderstood the reduction proposal somehow.
I can quickly update the patch to make reduction_impl constructor to accept 'T *ReduVarPtr' instead of 'T &ReduVar'.

Later, if there is the agreement regarding passing reference to Var, I can add 2 more constructors (from the current patch) to reduction (to support variant 2 below).

```
//Variant1
queue Q;
T* ReduVarPtr = malloc_device(sizeof(T), q);
*ReduVarPtr = Identity;
Q.submit([&](handler &CGH) {
  auto Redu = intel::reduction(ReduVarPtr, Identity, BOp);
  ...
});
//Variant2
queue Q;
T ReduVar = Identity;
Q.submit([&](handler &CGH) {
  auto Redu = intel::reduction(ReduVar, Identity, BOp);
  ...
});
```

BTW, fix me if I am wrong here: handler::copy(accessor, ptr) does not require 'ptr' to be USM pointer, right? If So, then in the current/proposed patch USM'ness is just not required at all.

Comment (v-klochkov) [reply sycl/test/reduction/reduction_usm.cpp]: John, I did the additional fix. Please see the 2nd commit in this PR.
Thank you!

**Comment by v-klochkov** [comment]
The 2nd commit in this PR caused a new fail on the new LIT test reduction_usm.cpp.
I don't see why it fails on GPU for sycl-ubu-x64-pr buildbot. I cannot reproduce that fail locally.

One of changes in the 2nd patch that potentially could break something is the explicit memory allocation/free.
So, I add the 3rd commit with a proactive fix for one of places that looked suspicious to me. Perhaps, it fixes the issue.

File: sycl/test/reduction/reduction_usm.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
+// UNSUPPORTED: cuda
+// Reductions use work-group builtins not yet supported by CUDA.
+
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUNx: env SYCL_DEVICE_TYPE=HOST %t.out
```
Comment (vladimirlaz) [inline sycl/test/reduction/reduction_usm.cpp:5]: Why do we have RUNx here.  If host is not suppoerted add corresponding feature to UNSUPPORTED section with proper comment.

Comment (v-klochkov) [reply sycl/test/reduction/reduction_usm.cpp:5]: I see many other 'RUNx' in other SYCL LIT tests, that is why I thought it is Ok to have it.

Vladimir, Can you please give one good example where it is done properly (with UNSUPPORTED keyword)
Did you really mean to write something like 'UNSUPPORTED: host', or just delete the line with RUNx and write a comment section. Something like:
// TODO: enable this test on HOST when HOST gets support of intel::reduce() and barrier() features.

File: sycl/include/CL/sycl/handler.hpp [hunk 4]
```diff
@@ -865,8 +852,22 @@ class __SYCL_EXPORT handler {
   detail::enable_if_t<Reduction::accessor_mode == access::mode::read_write &&
                       Reduction::has_fast_atomics>
   parallel_for(nd_range<Dims> Range, Reduction &Redu, KernelType KernelFunc) {
-    intel::detail::reduCGFunc<KernelName>(*this, KernelFunc, Range, Redu,
-                                          Redu.MAcc);
+    if (Reduction::is_usm)
+      Redu.associateWithHandler(*this);
+    auto QueueCopy = MQueue;
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/handler.hpp:857]: Please, do not use auto in such cases.

File: sycl/test/reduction/reduction_usm.cpp (hunk 2 above)
Comment (Pennycook) [reply sycl/test/reduction/reduction_usm.cpp]: I'm not sure about this.  If `handler::copy(accessor, ptr)` runs on the host, it won't be able to copy from a device pointer into an accessor.

@jbrodman: What is the right way to copy from a USM pointer into an accessor?

File: sycl/include/CL/sycl/handler.hpp (hunk 4 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:857]: I will fix this with 'shared_ptr_class<detail::queue_impl> QueueCopy = MQueue;' even though I do not agree with you regarding 'auto'.
The original statement was 'auto MQueueCopy = MQueue', which can be read as 'Whatever MQueue is, make a copy of it.'

One of key good features of 'auto' is "If you change the type of 'MQueue' in one place, you would not have to also update all other places where it is used, i.e. lines like the line 857"

Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp]: Fixed in additional commit.

File: sycl/test/reduction/reduction_usm.cpp (hunk 3 above)
Comment (v-klochkov) [reply sycl/test/reduction/reduction_usm.cpp]: @vladimirlaz I did additional changes (4th commit in this PR) per your comment.

File: sycl/test/reduction/reduction_usm.cpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    return;
+
+  // Initialize.
+  T CorrectOut;
+  BinaryOperation BOp;
+
+  buffer<T, 1> InBuf(NWItems);
+  initInputData(InBuf, CorrectOut, Identity, BOp, NWItems);
+
+  T *ReduVarPtr = (T *)malloc_shared(sizeof(T), Dev, Q.get_context());
```
Comment (Pennycook) [inline sycl/test/reduction/reduction_usm.cpp:40]: I've realized that the best way to address my concern from https://github.com/intel/llvm/pull/1657#discussion_r422208619 is probably going to be to repeat this test for all of the different [types of USM pointer](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/USM.adoc#usm-allocations).  Reductions should work for memory allocated with `malloc_host`, `malloc_device` or `malloc_shared`.

Since we haven't decided yet what should happen with non-USM pointers, we should consider adding a [runtime check](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/USM.adoc#usm-allocations) to make sure that the user hasn't accidentally provided a non-USM pointer.

File: sycl/test/reduction/reduction_usm.cpp [hunk 6]
```diff
@@ -1,12 +1,16 @@
...
+// UNSUPPORTED: linux
+// TODO: Enable the test for Linux when CI uses GPU driver 20.06.15619 or newer.
+
 // RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
-// RUNx: env SYCL_DEVICE_TYPE=HOST %t.out
 // RUN: %CPU_RUN_PLACEHOLDER %t.out
 // RUN: %GPU_RUN_PLACEHOLDER %t.out
 // RUN: %ACC_RUN_PLACEHOLDER %t.out
 
+// TODO: Enable the test for HOST when it supports intel::reduce() and barrier()
```
Comment (vladimirlaz) [inline sycl/test/reduction/reduction_usm.cpp:12]: Could you revert your change regarding HOST and UNSUPPORTED. The previous behavior was better because not the whole test is disabled for all targets. My comment should be treated as separate PR,

File: sycl/test/reduction/reduction_usm.cpp [hunk 7]
```diff
@@ -1,12 +1,16 @@
 // UNSUPPORTED: cuda
 // Reductions use work-group builtins not yet supported by CUDA.
 
+// UNSUPPORTED: linux
```
Comment (vladimirlaz) [inline sycl/test/reduction/reduction_usm.cpp:4]: For bugs it is better to use XFAIL instead of UNSUPPORTED because it will XPASS once driver upgrade happens and we do not forget to turn it on.

File: sycl/test/reduction/reduction_placeholder.cpp [hunk 8]
```diff
@@ -2,11 +2,12 @@
 // Reductions use work-group builtins not yet supported by CUDA.
 
 // RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
-// RUNx: env SYCL_DEVICE_TYPE=HOST %t.out
 // RUN: %CPU_RUN_PLACEHOLDER %t.out
 // RUN: %GPU_RUN_PLACEHOLDER %t.out
 // RUN: %ACC_RUN_PLACEHOLDER %t.out
 
+// TODO: Enable the test for HOST when it supports intel::reduce() and barrier()
+
```
Comment (vladimirlaz) [inline sycl/test/reduction/reduction_placeholder.cpp:10]: Could you revert your change regarding HOST and UNSUPPORTED. The previous behavior was better because not the whole test is disabled for all targets. My comment should be treated as separate PR.

File: sycl/test/reduction/reduction_usm.cpp (hunk 5 above)
Comment (v-klochkov) [reply sycl/test/reduction/reduction_usm.cpp]: @Pennycook , Your concern was regarding host memory, thus I added checks for USM pointers allocated via malloc_host() as well. Now alloc_shared/alloc_host memory is used/tested in reduction.
I don't see how reduction would work with var allocated via malloc_device(). That memory is assumed to be not accessible on HOST, right?

**Comment by v-klochkov** [comment]
@vladimirlaz , please see the changes in LIT tests. I removed 'UNSUPPORTED: linux' line because it blocked the test on Linux even for CPU where the test passed.
Also, I did not turn off the test on Linux-GPU because the cause of the fail is old GPU driver used in CI and which is going to be updated very soon.
Thank you.

File: sycl/test/reduction/reduction_usm.cpp (hunk 5 above)
Comment (Pennycook) [reply sycl/test/reduction/reduction_usm.cpp]: > I don't see how reduction would work with var allocated via malloc_device(). That memory is assumed to be not accessible on HOST, right?

This is the case I'm worried about, and why I think there should be a test with `malloc_device`.

I appreciate that it may not be clear from the reduction proposal, but the intent was that USM reductions would behave the same way as buffer/accessor reductions.  The result of a reduction should not be written immediately back to the host, but into the device memory specified by the USM pointer or accessor.  If the result is needed on the device for a subsequent kernel, no "extra" host synchronization is required; if the result is needed on the host, it's the user's responsibility to get it there (e.g. by creating a host accessor, using an explicit `memcpy` from a device pointer, or synchronizing the queue before reading from a shared pointer).

Since the result isn't guaranteed to be written back to the host, simply copying between any temporary space used by the reduction (on the device) and the device pointer initially specified by the user should be sufficient.  The host has to initiate the copy, but doesn't ever need to read the value itself.  I don't know whether you can achieve that with `handler::copy`, or if you'd need to use something else, hence the question in https://github.com/intel/llvm/pull/1657#discussion_r422208619.

Does that make sense?

Comment (v-klochkov) [reply sycl/test/reduction/reduction_usm.cpp]: @Pennycook , Thank you for the explanations. I hope I understood the idea and added check of USM memory allocated via malloc_device() to the LIT test.
Please see the 5th commit.

Comment (Pennycook) [reply sycl/test/reduction/reduction_usm.cpp]: Yes, this looks right to me.  Thanks, @v-klochkov!

File: sycl/include/CL/sycl/intel/reduction.hpp (hunk 1 above)
Comment (Pennycook) [reply sycl/include/CL/sycl/intel/reduction.hpp:492]: Ok, that makes sense.  Thanks for the explanation.

**Comment by v-klochkov** [comment]
After John's approval I simply re-based the patch to catch the new HEAD with fixed/updated CI. I did not do additional changes.

**Comment by bader** [comment]
> After John's approval I simply re-based the patch to catch the new HEAD with fixed/updated CI. I did not do additional changes.

Please, use "merge" instead of "rebase".

**Comment by bader** [comment]
I still see unresolved comments from Vlad and John. What is the plan to resolve them?

File: sycl/test/reduction/reduction_usm.cpp (hunk 2 above)
Comment (Pennycook) [reply sycl/test/reduction/reduction_usm.cpp]: I'm still slightly surprised that `CopyHandler.copy(Acc, Redu.getUSMPointer());` does the right thing here, but since the `malloc_device` test is passing I'll mark this resolved.

**Comment by v-klochkov** [comment]
> I still see unresolved comments from Vlad and John. What is the plan to resolve them?

All comments are resolved now.
